### PR TITLE
Adds a TempFile utility class to core's file utilities

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This is a feature PR that add a utility class to help with generating temp files with unique names
- By default, TempFile removes the file once it goes out of scope, but the constructor has a flag to keep the file
- It incorporates this `TempFile` class into several of Axom's unit tests.
- It also fixes a warning from Sina's tests about using an unsafe system call (`tmpnam`), and calls the recommended `mkstemp` instead.
   - Removing that warning was the impetus for adding the TempFile class.
- Along the way it also fixes some edge cases in Axom's `joinPath` file utility
- Resolves a subtask of #1467

### CODE REVIEW

File: src/axom/core/tests/utils_fileUtilities.hpp [hunk 1]
```diff
@@ -30,18 +32,40 @@ TEST(utils_fileUtilities, joinPath)
...
+  }
 
-  EXPECT_EQ(fullfile, fs::joinPath(fdirWithSlash, fname));
+  // test with a different separator
+  {
+    EXPECT_EQ("abc.def", fs::joinPath("abc", "def", "."));
+    EXPECT_EQ("abc.def", fs::joinPath("abc.", "def", "."));
+    EXPECT_EQ("abc.def", fs::joinPath("abc", ".def", "."));
+    EXPECT_EQ("abc.def", fs::joinPath("abc.", ".def", "."));
+  }
```
Comment (Arlie-Capps) [inline src/axom/core/tests/utils_fileUtilities.hpp:57]: Please add a joinPath test with the backslash, since that character is treated differently than most, and since Windows uses backslash.

File: src/axom/core/utilities/FileUtilities.hpp [hunk 2]
```diff
@@ -119,6 +121,97 @@ void getDirName(std::string& dir, const std::string& path);
...
+ * The TempFile class provides a convenient way to create and manage temporary files.
+ * It ensures that the file is deleted upon destruction, if specified.
+ * The class is non-copyable to prevent accidental duplication of file handles.
+ *
+ * \note The path to the temp file is likely different from the supplied \a file_name
+ */
+class TempFile
+{
+public:
+  /**
```
Comment (BradWhitlock) [inline src/axom/core/utilities/FileUtilities.hpp:136]: Maybe in the future, a default constructor that makes up a random alphanumeric filename out in $TMPDIR could be useful.

File: src/axom/quest/tests/quest_sampling_shaper.cpp [hunk 3]
```diff
@@ -62,27 +62,31 @@ const std::string proe_tet_fmt_str = R"(
 // Set the following to true for verbose output and for saving vis files
 constexpr bool very_verbose_output = false;
 
-/// RAII utility class to write a file at construction time and remove it
-/// once the instance is out of scope
+/// RAII utility class to write a file and remove it when the instance is out of scope
 class ScopedTemporaryFile
```
Comment (BradWhitlock) [inline src/axom/quest/tests/quest_sampling_shaper.cpp]: Why is this class needed when TempFile does the same thing?

File: src/axom/core/utilities/FileUtilities.hpp [hunk 4]
```diff
@@ -119,6 +121,97 @@ void getDirName(std::string& dir, const std::string& path);
...
+   * \brief Creates a tmp file using \a file_name and \a file_ext
```
Comment (rhornung67) [inline src/axom/core/utilities/FileUtilities.hpp]: Minor nit: Some comments use `tmp`, some use `temp`, and others use `temporary`.  I find this distracting. Please ignore if others don't.

File: src/axom/core/utilities/FileUtilities.hpp [hunk 5]
```diff
@@ -119,6 +121,97 @@ void getDirName(std::string& dir, const std::string& path);
...
+   * 
+   * \param file_name The name of the temporary file to create without extension
+   * \param file_ext An optional extension for the temp file
+   * \param delete_during_destruction If true (default), the file will be deleted when the TempFile object is destroyed.
```
Comment (rhornung67) [inline src/axom/core/utilities/FileUtilities.hpp]: Question: If the file is not deleted when the object is destroyed, and the file name is randomly generated, how does one clean it up?

File: src/axom/core/utilities/FileUtilities.hpp (hunk 4 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp]: Thanks -- I'll update it

File: src/axom/core/utilities/FileUtilities.hpp (hunk 5 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp]: I was mostly thinking this would be useful for short term debugging, e.g. if we need to check that the generated files match our expectations

File: src/axom/quest/tests/quest_sampling_shaper.cpp (hunk 3 above)
Comment (kennyweiss) [reply src/axom/quest/tests/quest_sampling_shaper.cpp]: I expected to remove it in favor of the new `TempFile`, but it had two nice features that were used dozens of times in this file and would have required a lot more changes -- 
1. Taking the string in the constructor and writing it out immediately
2. The `getFileContents` function which reads the file and returns its contents as a string

I'm not sure that either of them made sense to add to the new `TempFile` class directly, so I thought it'd be best to encapsulate it as a local small class in this file.

I'd love to hear suggestions on how to improve it though...

File: src/axom/core/utilities/FileUtilities.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp:136]: I think this effectively does this.
* On unix, there is a six digit string that gets generated by mkstemp.
* On windows, it takes the first three chars of the prefix and then add random chars.

I haven't tested this w/ an empty file_name prefix, but I think it would give the random filename you're looking for.
I'll add a test and post my findings.

File: src/axom/core/tests/utils_fileUtilities.hpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/core/tests/utils_fileUtilities.hpp:57]: Done

File: src/axom/core/utilities/FileUtilities.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp:136]: Actually, I already had that test!

Here's what we get on unix for the mapping from file_name and file_ext to the generated tmp file:
* ("", "") -> `$TMPDIR/oKOBd9`
* ("", ".json") - > `$TMPDIR/Mvgj3F.json`
* ("", "json") -> `$TMPDIR/GdOtkN.json`
* ("foo", "") ->`$TMPDIR/fooqN5Zd8`
* ("foo", ".json") ->`$TMPDIR/fooLC8LzV.json`
* ("foo", "json") ->`$TMPDIR/foo3LAglR.json`

And on windows:
* ("", "") -> `$TMPDIR/B2C1.tmp`
* ("", ".json") - > `$TMPDIR/B2C2.json`
* ("", "json") -> `$TMPDIR/B2C2.json`
* ("foo", "") ->`$TMPDIR/fooB2C4.tmp`
* ("foo", ".json") ->`$TMPDIR/fooB2C5.json`
* ("foo", "json") ->`$TMPDIR/fooB2C6.json`

where `$TMPDIR` was `/var/tmp/$user` on unix and `C:\Users\RUNNER~1\AppData\Local\Temp\` on Windows

File: src/axom/quest/tests/quest_pro_e_reader.cpp [hunk 6]
```diff
@@ -24,37 +27,37 @@ namespace
...
+  file.close();
 }
 
 } /* end anonymous namespace */
 
 //------------------------------------------------------------------------------
 TEST(quest_pro_e_reader, read_missing_file)
 {
-  const std::string INVALID_FILE = "foo.proe";
+  const std::string INVALID_FILE = "nonexistant_file.proe";
```
Comment (Arlie-Capps) [inline src/axom/quest/tests/quest_pro_e_reader.cpp]: spelling: "nonexistent_file"  (minor, of course)  : )

File: src/axom/quest/tests/quest_sampling_shaper.cpp (hunk 3 above)
Comment (Arlie-Capps) [reply src/axom/quest/tests/quest_sampling_shaper.cpp]: I could see features 1 and 2 staying in this small local class, but I also think they would fit comfortably in the new `TempFile` class.  Especially (2).

**Review by Arlie-Capps — APPROVED** [review]
Looks like a very useful addition.  Thank you for the tests.

File: src/axom/core/utilities/FileUtilities.hpp (hunk 4 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp]: Done

File: src/axom/quest/tests/quest_pro_e_reader.cpp (hunk 6 above)
Comment (kennyweiss) [reply src/axom/quest/tests/quest_pro_e_reader.cpp]: Done

File: src/axom/quest/tests/quest_sampling_shaper.cpp (hunk 3 above)
Comment (kennyweiss) [reply src/axom/quest/tests/quest_sampling_shaper.cpp]: Thanks -- I'll add the `getFileContents()` function to the main class

Comment (kennyweiss) [reply src/axom/quest/tests/quest_sampling_shaper.cpp]: Thanks for the suggestions!

I ended up adding both (1) and (2) to the `TempFile` class and removed the custom `ScopedTemporaryFile` class from the sampling shaping tests.

File: src/axom/core/utilities/FileUtilities.hpp (hunk 5 above)
Comment (kennyweiss) [reply src/axom/core/utilities/FileUtilities.hpp]: Since we're not expecting to use this flag that often, I moved it out of the constructor and added a `retain(bool)` function.

**Comment by kennyweiss** [comment]
Thanks for all the PR suggestions! I think they've improved the `TempFile` class and its usage
